src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -175,7 +165,29 @@ private void CloseInternal()
         private SecurityStatusPal EncryptData(ReadOnlyMemory<byte> buffer, ref byte[] outBuffer, out int outSize)
         {
             ThrowIfExceptionalOrNotAuthenticated();
-            return _context!.Encrypt(buffer, ref outBuffer, out outSize);
+
+            TaskCompletionSource<bool>? waiter = null;
+            while (true)
+            {
+                if (waiter != null)
+                {
+                    waiter.Task.Wait();
+                    _handshakeWaiter = null;
```

Comment:
What prevents this from stomping on a write to _handshakeWaiter from another thread?

Reply:
I probably don't understand the comment. There is always only one read and one write at max. That is enforced via atomic _nestedRead and _nestedWrite.  Please take a look at updated processing.